File: BankKata/BankCore/Services/StatementPrinter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography.X509Certificates;
using BankCore.Models;

namespace BankCore.Services
{
    public class StatementPrinter
    {
        private readonly TestConsole _console;
        private readonly List<string> _transactionStrings;

        public StatementPrinter(TestConsole console)
        {
            _console = console;
            _transactionStrings = new List<String>();
        }

        public List<string> TransactionStrings
        {
            get { return new List<string>(_transactionStrings.AsReadOnly()); }
        }

        public virtual void Print(List<Transaction> initialTransactions)
        {
            _console.PrintLine("DATE | AMOUNT | BALANCE");

            var tempBalance = 0;

            foreach (var transaction in initialTransactions.OrderBy(T=>Convert.ToDateTime(T.Date)))
            {
                _transactionStrings.Add(transaction.Date + " | " + transaction.Amount.ToString("F2") + " | " + (tempBalance += transaction.Amount).ToString("F2"));
            }

            _transactionStrings.Reverse();

            foreach (var transactionString in _transactionStrings)
            {
                _console.PrintLine(transactionString);
            }
        }
    }
}

File: BankKata/BankCoreVBNetEdition/StatementPrinter.vb
Public Class StatementPrinter

    Private ReadOnly _console As BaseConsole
    Private ReadOnly _transactionStrings As List(Of String)

    Public Sub New(console As BaseConsole)

        _console = console
        _transactionStrings = New List(Of String)()

    End Sub

    Public ReadOnly Property TransactionStrings As List(Of String)
        Get
            Return New List(Of String)(_transactionStrings.AsReadOnly())
        End Get
    End Property

    Public Overridable Sub Print(initialTransactions As List(Of Transaction))
        _console.PrintLine("DATE | AMOUNT | BALANCE")

        Dim runningBalance As Integer

        For Each transaction As Transaction In initialTransactions.OrderBy(Function(T) Convert.ToDateTime(T.Date))
            runningBalance += transaction.Amount
            _transactionStrings.Add(transaction.Date + " | " + transaction.Amount.ToString("F2") + " | " + runningBalance.ToString("F2"))
        Next

        _transactionStrings.Reverse()

        For Each transactionString As String In _transactionStrings
            _console.PrintLine(transactionString)
        Next

    End Sub
End Class